Code/GraphMol/Sgroup.h

```diff
@@ -0,0 +1,215 @@
+//
+//
+//  Copyright (C) 2002-2018 Greg Landrum and T5 Informatics GmbH
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+/*! \file Sgroup.h
+
+  \brief Defines the SGroup class
+
+*/
+#include <RDGeneral/export.h>
+#ifndef _RD_SGROUP_H
+#define _RD_SGROUP_H
+
+#include <unordered_map>
+
+#include <Geometry/point.h>
+#include <RDGeneral/types.h>
+#include <boost/smart_ptr.hpp>
+
+namespace RDKit {
+class ROMol;
+class Bond;
+class Atom;
+
+//! used to indicate errors from incorrect sgroup access
+class RDKIT_GRAPHMOL_EXPORT SGroupException : public std::runtime_error {
+ public:
+  //! construct with an error message
+  SGroupException(const char *msg) : std::runtime_error(msg){};
+  //! construct with an error message
+  SGroupException(const std::string &msg) : std::runtime_error(msg){};
+};
+
+class RDKIT_GRAPHMOL_EXPORT SGroup {
+ public:
+  enum class BondType {
+    XBOND,  // External
+    CBOND,  // Internal
+  };
+
+  typedef std::array<RDGeom::Point3D, 3> Bracket;
+
+  struct AttachPoint {
+    Atom *aAtom;
+    Atom *lvAtom;
+    std::string id;
+  };
+
+  struct CState {
+    Bond *bond;
+    boost::shared_ptr<RDGeom::Point3D> vector;
+  };
+
+  typedef std::vector<Atom *> ATOM_PTR_VECT;
+  typedef ATOM_PTR_VECT::iterator ATOM_PTR_VECT_I;
+  typedef ATOM_PTR_VECT::const_iterator ATOM_PTR_VECT_CI;
+  typedef std::vector<Bond *> BOND_PTR_VECT;
+  typedef BOND_PTR_VECT::iterator BOND_PTR_VECT_I;
+  typedef BOND_PTR_VECT::const_iterator BOND_PTR_VECT_CI;
+
+  //! No default constructor
+  SGroup() = delete;
+
+  //! Constructor. Ownsership is only set on this side of the
+  //! relationship: mol->addSGroup(sgroup) still needs to be
+  //! called to get ownership on the other side.
+  SGroup(ROMol *owning_mol, const std::string &_type);
+
+  //! Copy Constructor: initialize from a second sgroup.
+  SGroup(const SGroup &other) = default;
+
+  //! Destructor
+  ~SGroup(){};
+
+  //! Get the molecule that owns this conformation
+  ROMol &getOwningMol() const { return *dp_mol; }
+
+  //! get the ID of this sgroup
+  inline unsigned int getId() const { return d_id; }
+
+  //! get the COMPNO of this sgroup
```

Comment:
If we switch SGroups to inherit from `RDProps` (as suggested in another comment), this very Mol-file specific attribute could be handled using the property interface instead of cluttering the public API.

Reply:
Right. There's also other properties like the "compno", or the SGroup type that we can store as RDProps.

Even more if we switch from using Atom* and Bond* to using indexes, so we can store them as std::vector<unsigned int> in RDProps. This will take some work, but would allow us to get rid of the "updateOwningMol" method, since we wouldn't need to update the pointers when making copies of mols.